submissions/zalevskyi/frogger-game/app.js

```diff
@@ -0,0 +1,340 @@
+            c.timeToAppear=false
+            c.timeToDisappear=false
+        })
+        for (let direction in this.moveStatus) this.moveStatus[direction] = false
+        gameIsOver = true
+        gameIsOn = false
+    }
+    renderRules() {
+        ctx.fillStyle = 'whitesmoke'
+        ctx.fillRect(50, 155, 405, 200)
+        ctx.font = '25px san-serif'
+        ctx.fillStyle = 'navy'
+        ctx.textAlign = 'center'
+        ctx.fillText('Gain one point for reaching water', 252, 185)
+        ctx.fillText('Collect stars for additional points', 252, 215)
+        ctx.fillText('Use arrows to move', 252, 245)
+        ctx.fillStyle = 'red'
+        ctx.fillText('Game ends when a bug hits you', 252, 275)
+        ctx.fillStyle = 'green'
+        ctx.fillText('Select character with arrows', 252, 305)
+        ctx.fillText('Or press Space to start', 252, 335)
+    }
+    renderSkinSelection() {
+        ctx.drawImage(Resources.get(IMAGES['selector']), PLAYER_POSITION_START.x, PLAYER_POSITION_START.y)
+        super.render()
+    }
+    changeSkin(change) {
+        let skin_index = (IMAGES['player'].indexOf(this.image) + change) % IMAGES['player'].length
+        if (skin_index < 0) skin_index = IMAGES['player'].length + skin_index
+        this.image = IMAGES['player'][skin_index]
+    }
+    renderScore() {
+        ctx.textAlign = 'left'
+        ctx.font = '30px san-serif'
+        ctx.fillText(`Score: ${this.score}`, SCORE_POSITION.x, SCORE_POSITION.y)
+    }
+    renderGameEnd() {
+        ctx.fillStyle = 'whitesmoke'
+        ctx.fillRect(50, 410, 405, 110)
+        ctx.font = '25px san-serif'
+        ctx.fillStyle = 'navy'
+        ctx.textAlign = 'center'
+        ctx.fillText('Game is Over', 252, 440)
+        ctx.fillText(`Your score: ${this.score}`, 252, 470)
+        ctx.fillText('Press Space to continue', 252, 500)
+    }
+}
+//Returns array of new collision circles instances for an image
+function collisionCirclesFactory(image) {
+    const RESULT = []
+    switch(image) {
+        case 'images/char-boy.png':
+        case 'images/char-cat-girl.png':
+        case 'images/char-horn-girl.png':
+        case 'images/char-pink-girl.png':
+        case 'images/char-princess-girl.png':
+            RESULT.push(new Circle(new Point2D(50, 94), 33))
+            RESULT.push(new Circle(new Point2D(50, 122), 17))
+            break
+        case 'images/enemy-bug.png':
+            RESULT.push(new Circle(new Point2D(30, 110), 30))
+            RESULT.push(new Circle(new Point2D(52, 110), 33))
+            RESULT.push(new Circle(new Point2D(75, 106), 23))
+            break
+        case 'images/Star.png':
+            RESULT.push(new Circle(new Point2D(51, 74), 9))
+            RESULT.push(new Circle(new Point2D(26, 94), 9))
+            RESULT.push(new Circle(new Point2D(73, 94), 9))
+            RESULT.push(new Circle(new Point2D(33, 122), 9))
+            RESULT.push(new Circle(new Point2D(63, 122), 9))
```

Comment:
Looks like a whole set of mysterious numbers. Aren't these calculable from some elementary sprite properties?
Not required to fix though.

Reply:
In this case it is really "magic" numbers inside the code. Sprites are not vector graphics generated by code, but ready to use pixel one. So its collision circles properties comes outside as specification for pixel image.